### PR TITLE
Errors, Granularity and Connection refactor

### PR DESCRIPTION
**Pull request by stewi1014**
I originally wanted to separate these changes into separate PRs for organisation's sake, but ended up settling on cleaning up the git history.

This worked for me a while ago, but is **currently untested** and examples are **not** updated to reflect changes. This is a work in progress and I'm hoping to get some opinions and input as I made these changes over a year ago. It definitely needs to be reviewed by someone who's more familiar with it.

Changes are
- Functions return errors
- OandaConnection has been refactored and renamed to Connection
- Granularity type has been added
- GetTimeToCandles and GetTimeFromCandles have been added
- go.mod and go.sum have been added

Unfortunately adding errors will require users of the library to make changes to their code. Adding go modules support and the tag 'v1.0.0' to before this PR and 'v2.0.0' to after it might help mitigate this for users of go modules. It would need to be merged before this PR.

### CODE REVIEW

**Comment by stewi1014** [comment]
@habuka036 Would you be available to help test this?

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	return apiErr
+}
+
+// APIError is returned when the Oanda server responds with an error
+//
+// Message is the returned error message from the server if possible to unmarshal,
+// otherwise it is simply the entire body of the responce
+type APIError struct {
+	Request  *http.Request
+	Responce *http.Response
```
Comment (spekulant) [inline errors.go]: Please fix the typos, there are a few (about 15) of the "responce -> response" ones in this PR.
```suggestion
	Response *http.Response
```

**Review by spekulant — COMMENTED** [review]
Just my two cents, not getting into the Maintainers way here - great work! I like that you introduced the golang-specific "line of sight" approach in a few places like in the `NewConnection` function. The new error handling you introduced is something that I believe everyone will benefit from, cool stuff! Finally, I really like that you made Granularity a first-class citizen here, it'll be much easier to work with than passing raw strings around. Most of the code looks good to me bar one comment I made below. I can't wait to see those new features in master :)
PS. I think this PR would get reviewed much sooner by the respectful Maintainers if you split it into a few smaller ones.

File: goanda.go [hunk 2]
```diff
@@ -2,115 +2,164 @@ package goanda
...
-type OandaConnection struct {
-	hostname       string
-	port           int
-	ssl            bool
-	token          string
-	accountID      string
-	DatetimeFormat string
-	headers        *Headers
+// NewConnection creates a new connection
+// This function calls Connection.CheckConfig(), returning any errors
```
Comment (stewi1014) [inline goanda.go]: Reminder to myself: Connection.CheckConfig -> Connection.CheckConnection

File: errors.go (hunk 1 above)
Comment (stewi1014) [reply errors.go]: Thanks for the input! You're absolutely right, this PR is quite large. I didn't want to dive too deep into refactoring it for fear of breaking something without realising.

I've fixed the typos. Thanks for pointing them out.

**Comment by stewi1014** [comment]
So, it's been a while.

I'm going to mark this as ready, and hope someone comes to review and merge. This PR was, to the best of my ability, working and ready to merge the last time I developed with it.

I am happy to help resolve any issues, but no longer use this library. It seems as though the original creator hasn't been active here either. Hopefully anyone coming here wanting to use the library can get some use out of it, and that my PR can help a little.

**Comment by AwolDes** [comment]
Sorry about the late merge - I have just turned on notifications for this project. Thanks @stewi1014 !

**Pull request merged**